### PR TITLE
Journalist can edit and delete articles

### PR DESCRIPTION
**Pull request by perhegg**
#### Description
As a Journalist,
In order to change articles I have written
I would like to be able to edit or delete articles

Changes proposed in this pull request:

* Create namespace for Journalist
* Add functionality for edit and delete articles
* Add relationship between user and article
#### What I have learned working on this feature: 
I have learnt how namespacing works and how to use form_with forms.

### CODE REVIEW

File: app/views/journalist/articles/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    <p>
+        <%= form.submit %>
+    </p>
+
+    <p> 
+        <%= link_to 'Delete', 
+            journalist_article_path(@article), 
+                method: :delete,
+                data: { confirm: 'Are you sure?' } 
+                    %>
```
Comment (Joberg95) [inline app/views/journalist/articles/edit.html.erb]: format this file

File: features/step_definitions/assertions_steps.rb [hunk 2]
```diff
@@ -2,10 +2,15 @@
     expect(page).to have_content content
 end
 
-Then("I should be on the landing page") do
-    expect(current_path).to eq root_path 
+Then("I should be on the {string} page") do |page|
```
Comment (Joberg95) [inline features/step_definitions/assertions_steps.rb]: dont think we should use page as argument

File: app/views/journalist/articles/edit.html.erb (hunk 1 above)
Comment (perhegg) [reply app/views/journalist/articles/edit.html.erb]: done

File: features/step_definitions/assertions_steps.rb (hunk 2 above)
Comment (perhegg) [reply features/step_definitions/assertions_steps.rb]: done

**Review by aleciabenjamin — DISMISSED** [review]

File: app/controllers/journalist/articles_controller.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
+class Journalist::ArticlesController < ApplicationController
+    before_action :check_journalist, only: [:create, :new]
+    before_action :authenticate_user!, only: [:create, :new]
+
+    def new
```
Comment (oliverochman) [inline app/controllers/journalist/articles_controller.rb]: It makes no sense to display all articles on the `new` action.

File: app/controllers/journalist/articles_controller.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        @articles = Article.all
+    end
+
+    def show
+        @article = Article.find(params[:id])
+    end
+
+    def create
```
Comment (oliverochman) [inline app/controllers/journalist/articles_controller.rb]: why do you need to have a `create` action for this feature

File: features/step_definitions/assertions_steps.rb [hunk 5]
```diff
@@ -2,10 +2,11 @@
...
-end
-
 Then("I should not see {string}") do |content|
     expect(page).to have_no_content content
-end  
\ No newline at end of file
+end  
+
+Then("I should be on the landing page") do
+    visit root_path
```
Comment (oliverochman) [inline features/step_definitions/assertions_steps.rb]: `expect(current_path).to eq root_path` 
This is an assertion step

File: features/visitor_can_read_a_specific_article.feature [hunk 6]
```diff
@@ -5,11 +5,20 @@ Feature: Visitor can read a specific article
     I would like to be able to click on an article and have it displayed
 
     Background: 
+        Given the following categories exist
```
Comment (oliverochman) [inline features/visitor_can_read_a_specific_article.feature]: No need to create categories like this, take a look at the step definition for creating the articles

File: features/visitor_can_see_articles_on_index_page.feature [hunk 7]
```diff
@@ -5,7 +5,15 @@ Feature: List articles on the landing page
     I would like to be able to see the articles listed on a page
 
     Background: 
-         Given the following articles exist
+        Given the following categories exist
```
Comment (oliverochman) [inline features/visitor_can_see_articles_on_index_page.feature]: No need to create categories like this, take a look at the step definition for creating the articles

File: app/views/journalist/articles/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+<p>
+    <strong>Lead:</strong>
+    <%= @article.lead %>
+</p>
+ 
+<p>
+    <strong>Text:</strong>
+    <%= @article.content %>
+</p>
+<%= link_to "Edit", edit_journalist_article_path %>
```
Comment (oliverochman) [inline app/views/journalist/articles/show.html.erb:15]: should all journalists be able to edit and delete all articles?

File: app/views/journalist/articles/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,18 @@
+<%= form_with scope: :article, url: journalist_articles_path, local: true do |form| %>
```
Comment (oliverochman) [inline app/views/journalist/articles/new.html.erb]: why is this added in this PR?

File: app/controllers/journalist/articles_controller.rb (hunk 3 above)
Comment (oliverochman) [reply app/controllers/journalist/articles_controller.rb]: And shouldn't a journalist be able to only see his own articles?

File: app/controllers/journalist/articles_controller.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
+class Journalist::ArticlesController < ApplicationController
+    before_action :check_journalist, only: [:create, :new]
```
Comment (oliverochman) [inline app/controllers/journalist/articles_controller.rb]: I would remove the `only: [:create, :new]`, visitors should not be able to run any of this controllers actions

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/controllers/journalist/articles_controller.rb (hunk 4 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb]: the create feature is already merged

File: features/journalist_can_edit_and_delete_articles.feature [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    Background: 
+        Given the following categories exist
+            | name                  | 
+            | Breaking News         | 
+            | Politics              | 
+            | Sport                 | 
+            | Lifestyle             | 
+            | Health                | 
+
+        Given the following articles exist
```
Comment (oliverochman) [inline features/journalist_can_edit_and_delete_articles.feature]: Refactor this whole feature file. Modify the background

**Review by oliverochman — CHANGES_REQUESTED** [review]

**Comment by oliverochman** [comment]
And please git pong

File: app/controllers/journalist/articles_controller.rb (hunk 10 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb]: done

File: app/views/journalist/articles/new.html.erb (hunk 9 above)
Comment (perhegg) [reply app/views/journalist/articles/new.html.erb]: it from an merged pr. i tried to solve the form_with issue and deleted and added links. maybe thats why it shows as added.

File: app/views/journalist/articles/show.html.erb (hunk 8 above)
Comment (perhegg) [reply app/views/journalist/articles/show.html.erb:15]: thats a later user story

File: app/controllers/journalist/articles_controller.rb (hunk 3 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb]: i'll move it to index. journalist seeing only his own is a later user story.

File: features/step_definitions/assertions_steps.rb (hunk 5 above)
Comment (perhegg) [reply features/step_definitions/assertions_steps.rb]: hm.. isn't it already in that folder..?

File: features/visitor_can_read_a_specific_article.feature (hunk 6 above)
Comment (NiciP) [reply features/visitor_can_read_a_specific_article.feature]: Will you please clarify for us... we thought this is how it is done in the feature file?

Comment (perhegg) [reply features/visitor_can_read_a_specific_article.feature]: done

File: features/visitor_can_see_articles_on_index_page.feature (hunk 7 above)
Comment (perhegg) [reply features/visitor_can_see_articles_on_index_page.feature]: done

File: features/journalist_can_edit_and_delete_articles.feature (hunk 11 above)
Comment (perhegg) [reply features/journalist_can_edit_and_delete_articles.feature]: done

File: app/controllers/journalist/articles_controller.rb (hunk 3 above)
Comment (magnus-thor) [reply app/controllers/journalist/articles_controller.rb]: Per here is the user story
```
As a Journalist,
In order to change articles I have written
I would like to be able to edit or delete articles
```
If you look at line 2 it explicitly says he should be able to change his articles.

And you should not be fetching all articles in the new method if you want to stay restful.

File: app/controllers/journalist/articles_controller.rb (hunk 4 above)
Comment (magnus-thor) [reply app/controllers/journalist/articles_controller.rb]: That was merged in a previous pr, and you should not be fixing it in this story, create a chore to move the create functionality to the new controller.

File: app/models/article.rb [hunk 12]
```diff
@@ -1,4 +1,5 @@
 class Article < ApplicationRecord
+    has_one :user
```
Comment (magnus-thor) [inline app/models/article.rb]: Article should belong to a user

File: app/views/journalist/articles/new.html.erb (hunk 9 above)
Comment (magnus-thor) [reply app/views/journalist/articles/new.html.erb]: out of scope for this pr, add this to the chore of moving create to the new journalist controller.

File: app/views/journalist/articles/show.html.erb (hunk 8 above)
Comment (magnus-thor) [reply app/views/journalist/articles/show.html.erb:15]: If you look at my previous comment with the user story, the journalist should only be able to edit / delete his own articles.

File: features/journalist_can_edit_and_delete_articles.feature (hunk 11 above)
Comment (magnus-thor) [reply features/journalist_can_edit_and_delete_articles.feature]: The articles in the background need to associated with a journalist.

File: features/journalist_can_edit_and_delete_articles.feature [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            | Voted best mead recipe                    | Restaurant wins prize       | Restaurant wins prize for best mead in Sweden               | Lifestyle    |
+            | Ancient viking grave discovered           | Kids came across sword      | Kids come across sword protruding from the earth            | Breaking News|
+            | Drinking wine improves general health     | Drink wine today!           | Studies show that wine is good for your heart               | Health       |
+
+        And the following users exist
+            | email          | password | role       |
+            | jocke@craft.se | password | journalist |
+        And I am logged in as "jocke@craft.se"
+        And I visit the journalist page
+        Then I should see "Voted best mead recipe"
```
Comment (magnus-thor) [inline features/journalist_can_edit_and_delete_articles.feature]: Assertions do not belong in the background.

File: features/step_definitions/assertions_steps.rb (hunk 5 above)
Comment (magnus-thor) [reply features/step_definitions/assertions_steps.rb]: When you say `visit root_path` you are telling it to go to the root_path.
With `expect(current_path).to eq root_path` you are checking if you are on the root_path, there is  a big difference in those two things.

File: features/visitor_can_view_articles_in_categories.feature [hunk 14]
```diff
@@ -4,14 +4,6 @@ Feature: Visitor can view articles in categories
     I would like to be able to filter the articles displayed in categories
 
     Background:
```
Comment (magnus-thor) [inline features/visitor_can_view_articles_in_categories.feature:6]: Why are you removing the background step for creating the categories here?  it is not relevant for this PR.

**Review by magnus-thor — CHANGES_REQUESTED** [review]

File: app/controllers/journalist/articles_controller.rb (hunk 4 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb]: done

File: app/views/journalist/articles/new.html.erb (hunk 9 above)
Comment (perhegg) [reply app/views/journalist/articles/new.html.erb]: done

File: features/step_definitions/assertions_steps.rb (hunk 5 above)
Comment (perhegg) [reply features/step_definitions/assertions_steps.rb]: done

File: app/models/article.rb (hunk 12 above)
Comment (perhegg) [reply app/models/article.rb]: done

File: features/journalist_can_edit_and_delete_articles.feature (hunk 11 above)
Comment (perhegg) [reply features/journalist_can_edit_and_delete_articles.feature]: done

File: app/controllers/journalist/articles_controller.rb (hunk 3 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb]: done

File: app/views/journalist/articles/show.html.erb (hunk 8 above)
Comment (perhegg) [reply app/views/journalist/articles/show.html.erb:15]: done

File: app/controllers/journalist/articles_controller.rb [hunk 15]
```diff
@@ -0,0 +1,46 @@
+class Journalist::ArticlesController < ApplicationController
+    before_action :check_journalist
+    before_action :authenticate_user!
+    def index
```
Comment (magnus-thor) [inline app/controllers/journalist/articles_controller.rb]: The index action is not relevant for this feature.

File: app/controllers/journalist/articles_controller.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        @articles = Article.all.where(user_id: current_user.id)
+    end
+
+    def show
+        @article = Article.find(params[:id])
+    end
+    
+    def edit
```
Comment (magnus-thor) [inline app/controllers/journalist/articles_controller.rb:10]: Set a conditional that the user(journalist) can only access this article if he is the owner.

File: app/controllers/journalist/articles_controller.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+        @article = Article.find(params[:id])
+       
+        if @article.update(article_params)
+            redirect_to journalist_article_path, notice: 'Article was successfully updated.'
+        else
+            redirect_to @article, notice: 'Article was not updated.'
+        end
+    end
+
+    def destroy
```
Comment (magnus-thor) [inline app/controllers/journalist/articles_controller.rb:21]: Set a conditional that the user(journalist) can only access this article if he is the owner.

File: app/controllers/journalist/articles_controller.rb [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+        @article = Article.find(params[:id])
+        @article.destroy
+        
+        redirect_to journalist_articles_path, notice: 'Article was successfully deleted.'
```
Comment (magnus-thor) [inline app/controllers/journalist/articles_controller.rb]: This message will be rendered regardless of whether `@article.destroy` is successful or not. Set a conditional for either a successful or not successful `@article.destroy`

**Review by magnus-thor — CHANGES_REQUESTED** [review]

File: app/controllers/journalist/articles_controller.rb (hunk 18 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb]: done

File: app/controllers/journalist/articles_controller.rb (hunk 16 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb:10]: done

File: app/controllers/journalist/articles_controller.rb (hunk 17 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb:21]: done

File: app/controllers/journalist/articles_controller.rb (hunk 15 above)
Comment (perhegg) [reply app/controllers/journalist/articles_controller.rb]: done

**Review by magnus-thor — APPROVED** [review]
